### ISSUE TITLE
Requests which result in a 0DC do not have metric metadata

### ISSUE THREAD

**Issue by Stono** [istio/istio#25185]
**Bug description**
Another telemetry 2.0 gotcha.  If your requests get terminated by the source system (for example, they hit a timeout and the client gives up), these get reported (correctly) as a `0DC`.

In telemetry v2, such requests (which connected successfully, just terminated prematurely) do not have the destination metadata:
```
istio_requests_total{response_code="0",reporter="source",source_workload="security-service",source_workload_namespace="security-service",source_app="security-service",destination_workload="unknown",destination_workload_namespace="unknown",destination_app="unknown",destination_service="app.unified-registration-system.svc.cluster.local",destination_service_name="app.unified-registration-system.svc.cluster.local",destination_service_namespace="unknown",response_flags="DC"} 18
```

And then you can see in [these prometheus metrics](https://github.com/istio/istio/files/4866460/prometheus.log) we have:

```
istio_response_bytes_bucket{reporter="source",source_workload="security-service",source_workload_namespace="security-service",source_app="security-service",destination_workload="unified-registration-system",destination_workload_namespace="unified-registration-system",destination_app="unified-registration-system",destination_service="app.unified-registration-system.svc.cluster.local",destination_service_name="app",destination_service_namespace="unified-registration-system",le="0.5"} 0
```

and 

```
istio_response_bytes_bucket{reporter="source",source_workload="security-service",source_workload_namespace="security-service",source_app="security-service",destination_workload="unknown",destination_workload_namespace="unknown",destination_app="unknown",destination_service="app.unified-registration-system.svc.cluster.local",destination_service_name="app.unified-registration-system.svc.cluster.local",destination_service_namespace="unknown",le="0.5"} 18
```

notice the `destination workload` is unknown.

Effectively the same issue as https://github.com/istio/istio/issues/22154 I am guessing

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[x] Policies and Telemetry
[ ] Security
[ ] Test and Release
[x] User Experience
[ ] Developer Infrastructure

**Expected behavior**
Metrics representing client disconnections to be populated correctly 

**Steps to reproduce the bug**

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**
1.5.7

**How was Istio installed?**
Helm

**Environment where bug was observed (cloud vendor, OS, etc)**
GKE

**Comment by Stono** [comment]
cc @howardjohn @douglas-reid

**Comment by Stono** [comment]
:wave: @douglas-reid ping ping - this feels like more of an issue than the mirror metrics missing, this is actual miss-labeled telemetry in the case of DC (source service) disconnections

**Comment by douglas-reid** [comment]
I think this is related to the metadata for non-mesh services too, right?

We need a way to ensure proper metadata when exchange fails.

@kyessenov @gargnupur @mandarjog

**Comment by Stono** [comment]
They are mesh services, however a DC represents the calling service terminating before a response was received, for example during a client read timeout.

**Comment by nrjpoddar** [comment]
> They are mesh services, however a 0DC represents the calling service terminating before a response was received, for example during a client read timeout.
> 
> Likely the same fix for all the scenarios though as you say, fallback when metadata exchange doesn’t work.

@Stono if the client doesn't finish HTTP response header then the metadata exchange won't succeed which might be the issue here.

**Comment by Stono** [comment]
Yes that is the issue, and the fix is not relying on response data for meta info!

**Comment by bianpengyuan** [comment]
Dup of https://github.com/istio/istio/issues/20538. This will be addressed in 1.8
